application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java

```diff
@@ -65,26 +73,65 @@ public class DiagramMacroSolrMetadataExtractor implements SolrEntityMetadataExtr
     @Inject
     private EntityNameValidationConfiguration entityNameValidationConfiguration;
 
+    @Inject
+    private Logger logger;
+
     @Override
     public boolean extract(XWikiDocument document, SolrInputDocument solrDocument)
     {
 
-        List<Block> macroBlocks = document.getXDOM().getBlocks(new MacroBlockMatcher("diagram"), Block.Axes.CHILD);
-        if (macroBlocks != null && macroBlocks.size() > 0) {
+        XDOM xdom = document.getXDOM();
+        List<Block> macroBlocks = xdom.getBlocks(new MacroBlockMatcher("diagram"), Block.Axes.CHILD);
+        if (macroBlocks != null && macroBlocks.size() > 0 && !updateMacroReference(document, xdom, macroBlocks)) {
             List<DocumentReference> macroReferences = new ArrayList<>();
             for (Block macroBlock : macroBlocks) {
-                String referenceName =
-                    macroBlock.getParameter(REFERENCE) != null ? macroBlock.getParameter(REFERENCE) : "Diagram";
-
-                String transformedNamed = this.transformName(referenceName);
 
                 DocumentReference macroReference =
-                    explicitDocumentReferenceResolver.resolve(transformedNamed, document.getDocumentReference());
+                    explicitDocumentReferenceResolver.resolve(macroBlock.getParameter(REFERENCE),
+                        document.getDocumentReference());
                 macroReferences.add(macroReference);
             }
             return linkRegistry.registerBacklinks(solrDocument, macroReferences);
+        }
+        return false;
+    }
 
+    private boolean updateMacroReference(XWikiDocument document, XDOM xdom, List<Block> macroBlocks)
+    {
+        try {
 
+            XWikiContext context = contextProvider.get();
+            boolean modfied = false;
+            for (Block macroBlock : macroBlocks) {
+                String referenceName =
+                    macroBlock.getParameter(REFERENCE) != null ? macroBlock.getParameter(REFERENCE) : "Diagram";
+                // For backwards compatibility we check if the page already exists so we won't modify it.
+                DocumentReference macroReference =
+                    explicitDocumentReferenceResolver.resolve(referenceName, document.getDocumentReference());
```

Comment:
If we want to maintain backward compatibility, we have three cases:

    The reference does not follow the naming strategy, and the page is already created. In this case, we check if the page exists, and if it does, we retain the current reference.

    The reference does not follow the naming strategy, and the diagram was created in versions 1.20.7 or 1.20.8. In this case, the check for an existing page will fail, and the reference will be updated to follow the naming strategy. So instead of having {diagram reference="aaaa"/}, and transforming the naming strategy every time we load the macro, the macro call will become {diagram reference="bbbb"/} when reindexing the wiki, in the case of a naming strategy that transforms all "a"s into "b"s.

    The reference does not follow the naming strategy, and the page has not been created yet. In this case, when the wiki is reindexed, the reference will simply be updated to follow the naming strategy.